### PR TITLE
feat(vm-snapshots): implement vm snapshots details sidebar

### PR DESCRIPTION
**Pull request by tamazlykar**
Implements #1332

### CODE REVIEW

**Comment by wowshakhov** [comment]
Consider fixing https://github.com/bwsw/cloudstack-ui/blob/1b59e0c56345691a77f1222bce892df015d34672/src/app/snapshot/components/vm-snapshot-action-menu/vm-snapshot-action-menu.component.html#L1

File: src/app/snapshot/components/vm-snapshot-card-view/vm-snapshot-card-view.component.ts [hunk 1]
```diff
@@ -1,16 +1,27 @@
...
   templateUrl: './vm-snapshot-card-view.component.html',
   styleUrls: ['./vm-snapshot-card-view.component.scss'],
 })
-export class VmSnapshotCardViewComponent implements OnInit {
+export class VmSnapshotCardViewComponent {
   @Input()
   public item: VmSnapshotViewModel;
+  @Input()
+  public isSelected: (vmSnapshot: VmSnapshotViewModel) => boolean;
+  // tslint:disable-next-line:no-output-on-prefix
```
Comment (wowshakhov) [inline src/app/snapshot/components/vm-snapshot-card-view/vm-snapshot-card-view.component.ts]: Can we avoid this somehow?

File: src/app/snapshot/components/vm-snapshot-list-view/vm-snapshot-list-view.component.ts [hunk 2]
```diff
@@ -1,16 +1,27 @@
-import { Component, Input, OnInit } from '@angular/core';
+import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
```
Comment (wowshakhov) [inline src/app/snapshot/components/vm-snapshot-list-view/vm-snapshot-list-view.component.ts]: this looks identical to `vm-snapshot-card-view.component.ts`. Maybe we can fix this?

File: src/app/snapshot/components/vm-snapshot-sidebar/vm-snapshot-sidebar.component.html [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        [name]="'SNAPSHOT_PAGE.SIDEBAR.TYPE'"
+        [value]="vmSnapshot.type"
+      ></cs-parameters-pair>
+      <cs-parameters-pair
+        [name]="'SNAPSHOT_PAGE.SIDEBAR.CURRENT'"
+        [value]="vmSnapshot.current"
+      ></cs-parameters-pair>
+      <cs-parameters-pair
+        [name]="'SNAPSHOT_PAGE.SIDEBAR.PARENT'"
+        [value]="vmSnapshot.parentName ? vmSnapshot.parentName : '-'"
```
Comment (wowshakhov) [inline src/app/snapshot/components/vm-snapshot-sidebar/vm-snapshot-sidebar.component.html]: Why not use `vmSnapshot.parentName || '-'`?

File: src/app/snapshot/components/vm-snapshot-sidebar/vm-snapshot-sidebar.component.scss [hunk 4]
```diff
@@ -0,0 +1,3 @@
+.state-icon {
+  margin-right: 10px;
```
Comment (wowshakhov) [inline src/app/snapshot/components/vm-snapshot-sidebar/vm-snapshot-sidebar.component.scss]: Consider refactoring this into a separate component for state icon + text.

File: src/app/snapshot/containers/snapshots-page/snapshots-page.container.ts [hunk 5]
```diff
@@ -25,7 +25,7 @@ import { snapshotPageSelectors, vmSnapshotsSelectors } from '../../store';
     ></cs-snapshots-page>`,
 })
 export class SnapshotsPageContainerComponent implements OnInit, AfterViewInit {
-  readonly snapshots$ = this.store.pipe(select(fromSnapshots.selectFilteredSnapshots));
+  readonly snapshots$ = this.store.pipe(select(snapshotPageSelectors.getFilteredSnapshots));
```
Comment (wowshakhov) [inline src/app/snapshot/containers/snapshots-page/snapshots-page.container.ts]: Why did you rename select -> get? I think we should keep "select", because it's the default prefix in NgRx Entity:
`export const { selectIds, selectEntities, selectAll, selectTotal } = adapter.getSelectors(state);`

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 6]
```diff
@@ -1,11 +1,167 @@
...
+        vmState: vmEntities[vmSnapshot.virtualmachineid].state,
+        created: moment(vmSnapshot.created).toDate(),
+        type: vmSnapshot.type,
+        current: vmSnapshot.current,
+        parentName: vmSnapshot.parentName,
+      };
+      return viewModel;
+    }
+
+    return snapshot as Snapshot;
```
Comment (wowshakhov) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: Why do you only create ViewModel for VM snapshots? Volume snapshots have the `created` field as well.

File: src/app/snapshot/snapshots-page/snapshot-sidebar/snapshot-sidebar.container.ts [hunk 7]
```diff
@@ -1,30 +1,59 @@
...
+    select(snapshotPageSelectors.getViewMode),
+  );
   readonly volumes$ = this.store.pipe(select(fromVolumes.selectEntities));
-  readonly isLoading$ = this.store.pipe(select(fromSnapshots.isLoading));
+  readonly isLoading$ = combineLatest(
+    this.store.pipe(select(fromSnapshots.isLoading)),
+    this.store.pipe(select(vmSnapshotsSelectors.getIsLoading)),
+    // you need to wait for virtual machines to load for vmSnapshot selectors
+    this.store.pipe(select(fromVm.isLoading)),
+  ).pipe(map((loadings: boolean[]) => loadings.some(value => value)));
```
Comment (wowshakhov) [inline src/app/snapshot/snapshots-page/snapshot-sidebar/snapshot-sidebar.container.ts]: Why some? Don't you need everything to load? You should use `every` in this case.

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 8]
```diff
@@ -1,11 +1,167 @@
...
 export const getFilters = createSelector(getSnapshotPageState, (state): Filters => state.filters);
 
 export const getGroupings = createSelector(getSnapshotPageState, state => state.groupings);
+
+export const getSelectedSnapshot = createSelector(
+  getSelectedId,
+  vmSnapshotsSelectors.selectAll,
+  volumeSnapshotsSelectors.selectAll,
+  (selectedId, vmSnapsots, volumeSnapshots): Snapshot | VmSnapshot => {
+    const volumeSnapshot = volumeSnapshots.find(snap => snap.id === selectedId);
```
Comment (wowshakhov) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: Consider using `vmSnapsots.concat(volumeSnapshots).find(snap => snap.id === selectedId)`

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 9]
```diff
@@ -1,11 +1,167 @@
...
+const getFilteredVmSnapshots = createSelector(
+  vmSnapshotsSelectors.selectAll,
+  getFilters,
+  (snapshots, filters) => {
+    const filterByAccounts = (snapshot: VmSnapshot) => {
+      const filterEnabled = filters.accounts.length > 0;
+      if (!filterEnabled) {
+        return true;
+      }
+      return filters.accounts.findIndex(account => account === snapshot.account) >= 0;
```
Comment (wowshakhov) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: Use `includes` instead of `findIndex() >= 0`

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 10]
```diff
@@ -1,11 +1,167 @@
...
+      const end = moment(filters.date).endOf('day');
+      return moment(snapshot.created).isBetween(start, end);
+    };
+
+    const filterByVms = (snapshot: VmSnapshot) => {
+      const filterEnabled = filters.vmIds.length > 0;
+      if (!filterEnabled) {
+        return true;
+      }
+      return filters.vmIds.findIndex(id => id === snapshot.virtualmachineid) >= 0;
```
Comment (wowshakhov) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: Use `includes` instead of `findIndex() >= 0`

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 11]
```diff
@@ -1,11 +1,167 @@
...
+    );
+  },
+);
+
+export const getVmSnapshots = createSelector(
+  getFilteredVmSnapshots,
+  vmSelectors.selectEntities,
+  (snapshots, vmEntities): VmSnapshotViewModel[] => {
+    if (snapshots.length === 0) {
+      return [];
```
Comment (wowshakhov) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: You don't need to special case length = 0 here

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 12]
```diff
@@ -1,11 +1,167 @@
...
+      const vmName: string = vm ? vm.displayname : '';
+      const vmState: VmState = vm ? vm.state : undefined;
+      const viewModel: VmSnapshotViewModel = {
+        vmName,
+        vmState,
+        id: snapshot.id,
+        state: snapshot.state,
+        name: snapshot.displayname,
+        created: moment(snapshot.created).toDate(),
+        // necessary for grouping
```
Comment (wowshakhov) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: Specify what exactly is needed for grouping: account or account and domain.

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 13]
```diff
@@ -1,11 +1,167 @@
...
+    });
+  },
+);
+
+export const getFilteredSnapshots = createSelector(
+  volumeSnapshotsSelectors.selectAll,
+  getFilters,
+  (snapshots, filter) => {
+    const filterByTypes = (snapshot: Snapshot) =>
+      !filter.volumeSnapshotTypes.length ||
```
Comment (wowshakhov) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: Replace !!find with includes.

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 14]
```diff
@@ -1,11 +1,167 @@
...
+      moment(snapshot.created).isBetween(
+        moment(filter.date).startOf('day'),
+        moment(filter.date).endOf('day'),
+      );
+
+    const queryLower = filter.query && filter.query.toLowerCase();
+    const filterByQuery = (snapshot: Snapshot) => {
+      return (
+        !filter.query ||
+        snapshot.name.toLowerCase().indexOf(queryLower) > -1 ||
```
Comment (wowshakhov) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: Replace indexOf with includes

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 15]
```diff
@@ -1,11 +1,167 @@
...
+        (getSnapshotDescription(snapshot) &&
```
Comment (wowshakhov) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: Extract variable `const snapshotDescription = getSnapshotDescription(snapshot)`

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 16]
```diff
@@ -1,11 +1,167 @@
...
+          getSnapshotDescription(snapshot)
+            .toLowerCase()
+            .indexOf(queryLower) > -1)
```
Comment (wowshakhov) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: Replace indexOf with includes

File: src/app/snapshot/snapshots-page/snapshot-sidebar/snapshot-sidebar.container.ts [hunk 17]
```diff
@@ -1,30 +1,59 @@
...
+  );
+  readonly viewMode$: Observable<SnapshotPageViewMode> = this.store.pipe(
+    select(snapshotPageSelectors.getViewMode),
+  );
   readonly volumes$ = this.store.pipe(select(fromVolumes.selectEntities));
-  readonly isLoading$ = this.store.pipe(select(fromSnapshots.isLoading));
+  readonly isLoading$ = combineLatest(
+    this.store.pipe(select(fromSnapshots.isLoading)),
+    this.store.pipe(select(vmSnapshotsSelectors.getIsLoading)),
+    // you need to wait for virtual machines to load for vmSnapshot selectors
```
Comment (wowshakhov) [inline src/app/snapshot/snapshots-page/snapshot-sidebar/snapshot-sidebar.container.ts]: Maybe you should put the `fromVm.isLoading` check directly into vmSnapshotsSelectors.getIsLoading?

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 18]
```diff
@@ -1,11 +1,167 @@
...
+  (snapshot, vmEntities): Snapshot | VmSnapshotSidebarViewModel => {
+    if (getSnapshotType(snapshot) === SnapshotType.VM) {
+      const vmSnapshot: VmSnapshot = snapshot as VmSnapshot;
+      const viewModel: VmSnapshotSidebarViewModel = {
+        id: vmSnapshot.id,
+        state: vmSnapshot.state,
+        name: vmSnapshot.displayname,
+        description: vmSnapshot.description,
+        vmId: vmSnapshot.virtualmachineid,
+        vmName: vmEntities[vmSnapshot.virtualmachineid].displayname,
```
Comment (tamazlykar) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: check

**Comment by tamazlykar** [comment]
> Consider fixing in this PR:
> [cloudstack-ui/src/app/snapshot/components/vm-snapshot-action-menu/vm-snapshot-action-menu.component.html](https://github.com/bwsw/cloudstack-ui/blob/1b59e0c56345691a77f1222bce892df015d34672/src/app/snapshot/components/vm-snapshot-action-menu/vm-snapshot-action-menu.component.html#L1)
> 
> Line 1 in [1b59e0c](/bwsw/cloudstack-ui/commit/1b59e0c56345691a77f1222bce892df015d34672)
> 
>  <button mat-menu-item *ngIf="false" (click)="onCreateVolumeSnapshot()">

This function is not working on the server now, we enable it after it starts working.

File: src/app/snapshot/components/vm-snapshot-card-view/vm-snapshot-card-view.component.ts [hunk 19]
```diff
@@ -1,16 +1,23 @@
...
-export class VmSnapshotCardViewComponent implements OnInit {
+export class VmSnapshotCardViewComponent {
   @Input()
   public item: VmSnapshotViewModel;
+  @Input()
+  public isSelected: (vmSnapshot: VmSnapshotViewModel) => boolean;
+  @Output()
+  public clicked = new EventEmitter<VmSnapshotViewModel>();
+  @ViewChild(MatMenuTrigger)
+  public matMenuTrigger: MatMenuTrigger;
```
Comment (wowshakhov) [inline src/app/snapshot/components/vm-snapshot-card-view/vm-snapshot-card-view.component.ts]: Remove matMenuTrigger if it is not used

File: src/app/snapshot/components/vm-snapshot-list-view/vm-snapshot-list-view.component.ts [hunk 20]
```diff
@@ -1,16 +1,23 @@
...
-export class VmSnapshotListViewComponent implements OnInit {
+export class VmSnapshotListViewComponent {
   @Input()
   public item: VmSnapshotViewModel;
+  @Input()
+  public isSelected: (vmSnapshot: VmSnapshotViewModel) => boolean;
+  @Output()
+  public clicked = new EventEmitter<VmSnapshotViewModel>();
+  @ViewChild(MatMenuTrigger)
+  public matMenuTrigger: MatMenuTrigger;
```
Comment (wowshakhov) [inline src/app/snapshot/components/vm-snapshot-list-view/vm-snapshot-list-view.component.ts]: Remove matMenuTrigger if it is not used

File: src/app/snapshot/snapshots-page/snapshot-list-item/snapshot-card-item.component.ts [hunk 21]
```diff
@@ -23,9 +23,8 @@ export class SnapshotCardItemComponent extends SnapshotItemComponent {
   public isSelected: (snapshot: Snapshot) => boolean;
   @Input()
   public query: string;
-  // tslint:disable-next-line:no-output-on-prefix
   @Output()
-  public onClick = new EventEmitter<Snapshot>();
+  public clicked = new EventEmitter<Snapshot>();
   @ViewChild(MatMenuTrigger)
   public matMenuTrigger: MatMenuTrigger;
```
Comment (wowshakhov) [inline src/app/snapshot/snapshots-page/snapshot-list-item/snapshot-card-item.component.ts]: Remove matMenuTrigger if it is not used

File: src/app/snapshot/snapshots-page/snapshot-list-item/snapshot-item.component.ts [hunk 22]
```diff
@@ -11,7 +11,7 @@ export class SnapshotItemComponent {
   public virtualMachines: Dictionary<VirtualMachine>;
   public isSelected: (snapshot: Snapshot) => boolean;
   public query: string;
-  public onClick = new EventEmitter<Snapshot>();
+  public clicked = new EventEmitter<Snapshot>();
   public matMenuTrigger: MatMenuTrigger;
```
Comment (wowshakhov) [inline src/app/snapshot/snapshots-page/snapshot-list-item/snapshot-item.component.ts]: Remove matMenuTrigger if it is not used

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 23]
```diff
@@ -1,11 +1,159 @@
...
+      const vmName: string = vm ? vm.displayname : '';
+      const vmState: VmState = vm ? vm.state : undefined;
+      const viewModel: VmSnapshotViewModel = {
+        vmName,
+        vmState,
+        id: snapshot.id,
+        state: snapshot.state,
+        name: snapshot.displayname,
+        created: moment(snapshot.created).toDate(),
+        // account and domain necessary for grouping
```
Comment (wowshakhov) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: account and domain **are** necessary for grouping

File: src/app/snapshot/store/snapshot-page.selectors.ts [hunk 24]
```diff
@@ -1,11 +1,159 @@
...
+export const getFilteredSnapshots = createSelector(
+  volumeSnapshotsSelectors.selectAll,
+  getFilters,
+  (snapshots, filter) => {
+    const filterByTypes = (snapshot: Snapshot) =>
+      !filter.volumeSnapshotTypes.length ||
+      filter.volumeSnapshotTypes.includes(snapshot.snapshottype);
+
+    const filterByAccount = (snapshot: Snapshot) =>
+      !filter.accounts.length || !!filter.accounts.find(id => id === snapshot.account);
```
Comment (wowshakhov) [inline src/app/snapshot/store/snapshot-page.selectors.ts]: Replace !!find by includes

File: src/app/volume/volume-item/volume-item.component.ts [hunk 25]
```diff
@@ -46,10 +46,7 @@ export class VolumeItemComponent extends VolumeItem implements OnInit, OnChanges
     return stateTranslations[this.item.state.toUpperCase()];
   }
 
-  public handleClick(e: MouseEvent): void {
-    e.stopPropagation();
-    if (!this.matMenuTrigger.menuOpen) {
-      this.onClick.emit(this.item);
-    }
+  public handleClick(): void {
```
Comment (wowshakhov) [inline src/app/volume/volume-item/volume-item.component.ts]: Remove matMenuTrigger in this component if it's not used

File: src/app/account/account/account-item.component.ts [hunk 26]
```diff
@@ -19,11 +19,8 @@ export class AccountItemComponent {
 
   constructor(protected authService: AuthService) {}
 
-  public handleClick(e: MouseEvent): void {
-    e.stopPropagation();
-    if (!this.matMenuTrigger || !this.matMenuTrigger.menuOpen) {
-      this.onClick.emit(this.item);
-    }
+  public handleClick(): void {
```
Comment (wowshakhov) [inline src/app/account/account/account-item.component.ts]: Remove matMenuTrigger in this component if it's not used.

File: src/app/security-group/sg-list-item/security-group-list-item.component.ts [hunk 27]
```diff
@@ -34,11 +34,8 @@ export class SecurityGroupListItemComponent implements OnChanges {
     }
   }
 
-  public handleClick(e: MouseEvent): void {
-    e.stopPropagation();
-    if (!this.matMenuTrigger.menuOpen) {
-      this.onClick.emit(this.item);
-    }
+  public handleClick(): void {
```
Comment (wowshakhov) [inline src/app/security-group/sg-list-item/security-group-list-item.component.ts]: Remove matMenuTrigger in this component if it's not used.
